packages/editor/src/components/post-saved-state/index.js

```diff
@@ -2,175 +2,164 @@
-import { displayShortcut } from '@wordpress/keycodes';
-import { withSafeTimeout, compose } from '@wordpress/compose';
-import { withViewportMatch } from '@wordpress/viewport';
+import { usePrevious, useViewportMatch } from '@wordpress/compose';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { useEffect, useState } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
 import { Icon, check, cloud, cloudUpload } from '@wordpress/icons';
+import { displayShortcut } from '@wordpress/keycodes';
 
 /**
  * Internal dependencies
  */
 import PostSwitchToDraftButton from '../post-switch-to-draft-button';
 
 /**
- * Component showing whether the post is saved or not and displaying save links.
+ * Component showing whether the post is saved or not and providing save
+ * buttons.
  *
- * @param   {Object}    Props Component Props.
+ * @param {Object}   props               Component props.
+ * @param {?boolean} props.forceIsDirty  Whether to force the post to be marked
+ *                                       as dirty.
+ * @param {?boolean} props.forceIsSaving Whether to force the post to be marked
+ *                                       as being saved.
+ *
+ * @return {import('@wordpress/element').WPComponent} The component.
  */
-export class PostSavedState extends Component {
-	constructor() {
-		super( ...arguments );
-		this.state = {
-			forceSavedMessage: false,
-		};
-	}
+export default function PostSavedState( { forceIsDirty, forceIsSaving } ) {
+	const [ forceSavedMessage, setForceSavedMessage ] = useState( false );
+
+	const isLargeViewport = useViewportMatch( 'small' );
+
+	const {
+		isAutosaving,
+		isDirty,
+		isNew,
+		isPending,
+		isPublished,
+		isSaveable,
+		isSaving,
+		isScheduled,
+		post,
+	} = useSelect(
+		( select ) => {
+			const {
+				isEditedPostNew,
+				isCurrentPostPublished,
+				isCurrentPostScheduled,
+				isEditedPostDirty,
+				isSavingPost,
+				isEditedPostSaveable,
+				getCurrentPost,
+				isAutosavingPost,
+				getEditedPostAttribute,
+			} = select( 'core/editor' );
 
-	componentDidUpdate( prevProps ) {
-		if ( prevProps.isSaving && ! this.props.isSaving ) {
-			this.setState( { forceSavedMessage: true } );
-			this.props.setTimeout( () => {
-				this.setState( { forceSavedMessage: false } );
+			return {
+				isAutosaving: isAutosavingPost(),
+				isDirty: forceIsDirty || isEditedPostDirty(),
+				isNew: isEditedPostNew(),
+				isPending: 'pending' === getEditedPostAttribute( 'status' ),
+				isPublished: isCurrentPostPublished(),
+				isSaving: forceIsSaving || isSavingPost(),
+				isSaveable: isEditedPostSaveable(),
+				isScheduled: isCurrentPostScheduled(),
+				post: getCurrentPost(),
+			};
+		},
+		[ forceIsDirty, forceIsSaving ]
+	);
+
+	const { savePost } = useDispatch( 'core/editor' );
+
+	const wasSaving = usePrevious( isSaving );
+
+	useEffect( () => {
+		let timeoutId;
+
+		if ( wasSaving && ! isSaving ) {
```

Comment:
I don't think we need wasSaving, if isSaving is true we can assume the previous value was false since this effect triggers only on change (and mount but I assume it's false on mount)

Reply:
I'm not checking if `isSaving` is true here. I'm checking to see if it _was_ true and has just become false.

If it's false on mount, then that means if I was only checking for `! isSaving`, the "Saved" state would appear on load, which wouldn't be correct. So I'm pretty sure that `wasSaving` is necessary here.